crates/precompile/src/bn128/substrate.rs

```diff
@@ -0,0 +1,144 @@
+use super::{FQ2_LEN, FQ_LEN, G1_LEN};
+use crate::PrecompileError;
+use bn::{AffineG1, AffineG2, Fq, Fq2, Group, Gt, G1, G2};
+
+/// Reads a single `Fq` field element from the input slice.
+///
+/// Takes a byte slice and attempts to interpret the first 32 bytes as an
+/// elliptic curve field element. Returns an error if the bytes do not form
+/// a valid field element.
+///
+/// # Panics
+///
+/// Panics if the input is not at least 32 bytes long.
+#[inline]
+fn read_fq(input: &[u8]) -> Result<Fq, PrecompileError> {
+    Fq::from_slice(&input[..FQ_LEN]).map_err(|_| PrecompileError::Bn128FieldPointNotAMember)
+}
+/// Reads a Fq2 (quadratic extension field element) from the input slice.
+///
+/// Parses two consecutive Fq field elements as the real and imaginary parts
+/// of an Fq2 element.
+/// The second component is parsed before the first, ie if a we represent an
+/// element in Fq2 as (x,y) -- `y` is parsed before `x`
+///
+/// # Panics
+///
+/// Panics if the input is not at least 64 bytes long.
+#[inline]
+fn read_fq2(input: &[u8]) -> Result<Fq2, PrecompileError> {
+    let y = read_fq(&input[..FQ_LEN])?;
+    let x = read_fq(&input[FQ_LEN..2 * FQ_LEN])?;
+    Ok(Fq2::new(x, y))
+}
+
+/// Creates a new `G1` point from the given `x` and `y` coordinates.
+///
+/// Constructs a point on the G1 curve from its affine coordinates.
+///
+/// Note: The point at infinity which is represented as (0,0) is
+/// handled specifically because `AffineG1` is not capable of
+/// representing such a point.
+/// In particular, when we convert from `AffineG1` to `G1`, the point
+/// will be (0,0,1) instead of (0,1,0)
+#[inline]
+fn new_g1_point(px: Fq, py: Fq) -> Result<G1, PrecompileError> {
+    if px == Fq::zero() && py == Fq::zero() {
+        Ok(G1::zero())
+    } else {
+        AffineG1::new(px, py)
+            .map(Into::into)
+            .map_err(|_| PrecompileError::Bn128AffineGFailedToCreate)
+    }
+}
+
+/// Creates a new `G2` point from the given Fq2 coordinates.
+///
+/// G2 points in BN128 are defined over a quadratic extension field Fq2.
+/// This function takes two Fq2 elements representing the x and y coordinates
+/// and creates a G2 point.
+///
+/// Note: The point at infinity which is represented as (0,0) is
+/// handled specifically because `AffineG2` is not capable of
+/// representing such a point.
+/// In particular, when we convert from `AffineG2` to `G2`, the point
+/// will be (0,0,1) instead of (0,1,0)
+#[inline]
+fn new_g2_point(x: Fq2, y: Fq2) -> Result<G2, PrecompileError> {
+    // TODO : Check whether or not we need these zero checks
```

Comment:
This TODO was there before, we can remove it since the Note in the comments specify that we do need the check

Reply:
Makes sense lets remove it. That comment is probably there from few years ago :)

Reply:
```suggestion
```